File: go/internal_helpers/notify_client.go
// Package internalhelpers provides side-channel helpers for services that
// need to talk to octo-server (and octo-adjacent services) using the
// non-user-realm credential paths — the X-Internal-Token producer channel
// (see [NotifyClient]) and the X-*-Token consumer middleware (see
// [NewInternalTokenGinMiddleware] / [WrapInternalTokenNetHTTP]).
//
// These helpers are deliberately independent of the [Verifier] interface:
// they are service-to-service, not user-to-service, so they never build a
// [octoauth.Principal].
package internalhelpers

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"os"
	"time"
)

// DefaultNotifyTimeout is the fallback HTTP timeout when NotifyConfig.Timeout
// is zero. Ten seconds matches the internal-notify contract's SLA and the
// SMTP-style "fire and reasonably wait" pattern used by smart-summary and
// matter.
const DefaultNotifyTimeout = 10 * time.Second

// NotifyConfig configures a [NotifyClient].
//
// Required fields:
//   - BaseURL:     the target service base URL, e.g. https://octo-server:8090
//   - TokenEnvVar: the process env var name to read the shared-secret from,
//     e.g. "SUMMARY_NOTIFY_TOKEN"
//
// Optional fields fall back to safe defaults (see field docs).
type NotifyConfig struct {
	BaseURL     string
	TokenEnvVar string

	HTTPClient *http.Client  // default &http.Client{Timeout: Timeout}
	Timeout    time.Duration // default DefaultNotifyTimeout
	Logger     *slog.Logger  // default slog.Default()
	// NotifyPath overrides the default /v1/internal/notify endpoint path.
	// Leave empty to use the default; must start with "/" when set.
	NotifyPath string
}

// defaultNotifyPath is the octo-server internal-notify endpoint path.
const defaultNotifyPath = "/v1/internal/notify"

// NotifyClient posts service-authenticated notifications to octo-server via
// the shared-secret X-Internal-Token producer channel. Concurrent use is
// safe.
//
// The zero value is not usable; construct via [NewNotifyClient].
type NotifyClient struct {
	baseURL string
	path    string
	token   string
	client  *http.Client
	logger  *slog.Logger
}

// NewNotifyClient builds a [NotifyClient] from cfg. It reads the token from
// os.Getenv(cfg.TokenEnvVar) at construction time and returns an error when
// the value is empty — a "fail-closed at startup" posture that prevents the
// server from silently degrading into an unauthenticated caller.
//
// Errors:
//   - BaseURL empty
//   - TokenEnvVar empty
//   - env var not set or set to empty
func NewNotifyClient(cfg NotifyConfig) (*NotifyClient, error) {
	if cfg.BaseURL == "" {
		return nil, errors.New("octoauth/internal_helpers: NotifyConfig.BaseURL is required")
	}
	if cfg.TokenEnvVar == "" {
		return nil, errors.New("octoauth/internal_helpers: NotifyConfig.TokenEnvVar is required")
	}
	token := os.Getenv(cfg.TokenEnvVar)
	if token == "" {
		return nil, fmt.Errorf("octoauth/internal_helpers: env var %q is empty (fail-closed)", cfg.TokenEnvVar)
	}
	timeout := cfg.Timeout
	if timeout == 0 {
		timeout = DefaultNotifyTimeout
	}
	client := cfg.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: timeout}
	}
	logger := cfg.Logger
	if logger == nil {
		logger = slog.Default()
	}
	path := cfg.NotifyPath
	if path == "" {
		path = defaultNotifyPath
	}
	return &NotifyClient{
		baseURL: cfg.BaseURL,
		path:    path,
		token:   token,
		client:  client,
		logger:  logger,
	}, nil
}

// NotifyRequest is the payload sent to /v1/internal/notify. In v1 the shape
// is intentionally minimal (channel_id + message); callers who need to send
// a different shape can pre-marshal it into [NotifyRequest.RawBody] and
// leave the primary fields empty.
type NotifyRequest struct {
	ChannelID string `json:"channel_id,omitempty"`
	Message   string `json:"message,omitempty"`

	// RawBody, when non-nil, is sent verbatim as the request body and
	// the ChannelID/Message fields are ignored. Useful for callers whose
	// notify contract has already been agreed with the server and does
	// not match the v1 minimal shape.
	RawBody json.RawMessage `json:"-"`
}

// Send posts req to the notify endpoint. It is fire-and-forget in the sense
// that response bodies are drained but not parsed; only the HTTP status is
// consulted. A non-2xx status is returned as an error; the underlying body
// is NOT included so log-scrapers cannot enumerate server-side messages
// leaked via the caller's error path.
func (c *NotifyClient) Send(ctx context.Context, req NotifyRequest) error {
	var body []byte
	if len(req.RawBody) > 0 {
		body = req.RawBody
	} else {
		encoded, err := json.Marshal(req)
		if err != nil {
			return fmt.Errorf("octoauth/internal_helpers: encode notify request: %w", err)
		}
		body = encoded
	}

	url := c.baseURL + c.path
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("octoauth/internal_helpers: build notify request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("Accept", "application/json")
	httpReq.Header.Set("X-Internal-Token", c.token)

	resp, err := c.client.Do(httpReq)
	if err != nil {
		c.logger.Debug("octoauth/internal_helpers: notify transport error",
			slog.String("endpoint", c.path),
			slog.String("token_hint", maskToken(c.token)),
		)
		return fmt.Errorf("octoauth/internal_helpers: notify request failed: %w", err)
	}
	defer resp.Body.Close()
	// Drain the body so the connection can be reused.
	_, _ = io.Copy(io.Discard, resp.Body)

	c.logger.Debug("octoauth/internal_helpers: notify sent",
		slog.String("endpoint", c.path),
		slog.Int("status", resp.StatusCode),
		slog.String("token_hint", maskToken(c.token)),
	)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("octoauth/internal_helpers: notify failed: status %d", resp.StatusCode)
	}
	return nil
}

// maskToken returns a short prefix + "..." suitable for slog. Full-length
// masking is intentional — even 8 characters of an internal shared secret
// should not appear in application logs. Callers who need higher-fidelity
// diagnostics should use their metrics pipeline, not the logger.
func maskToken(t string) string {
	if len(t) < 8 {
		return "***"
	}
	return t[:8] + "..."
}

File: ts/src/internal-helpers/notify-client.ts
/**
 * internal-helpers/notify-client — X-Internal-Token producer channel.
 *
 * See design doc §9.1 and Go SDK's internal_helpers/notify_client.go.
 * Constructor reads the token from process.env at build time and throws if
 * empty — a "fail-closed at startup" posture matching Go's NewNotifyClient.
 */

import type { Logger } from '../config.js'
import { consoleLogger } from '../config.js'

/** DefaultNotifyTimeoutMs matches Go's DefaultNotifyTimeout (10s). */
export const DEFAULT_NOTIFY_TIMEOUT_MS = 10_000
const DEFAULT_NOTIFY_PATH = '/v1/internal/notify'

/** NotifyConfig configures a NotifyClient. */
export interface NotifyConfig {
  /** Required. Target service base URL, e.g. https://octo-server:8090. */
  baseUrl: string
  /** Required. Process env var to read the shared secret from. */
  tokenEnvVar: string
  /** Fetch implementation. Defaults to globalThis.fetch. */
  fetch?: typeof globalThis.fetch
  /** Per-request timeout ms. Defaults to 10_000. */
  timeoutMs?: number
  /** Logger. Defaults to consoleLogger. */
  logger?: Logger
  /** Optional path override. Must start with "/" when set. */
  notifyPath?: string
}

/**
 * NotifyRequest is the payload sent to /v1/internal/notify. In v1 the shape
 * is intentionally minimal (channel_id + message); callers whose contract
 * needs a different shape can pass `rawBody` and leave the primary fields
 * empty.
 */
export interface NotifyRequest {
  channelId?: string
  message?: string
  /** When set, this JSON string is sent verbatim; channelId/message ignored. */
  rawBody?: string
}

/**
 * NotifyClient posts service-authenticated notifications via the shared-secret
 * X-Internal-Token producer channel. Instances are safe for concurrent use.
 */
export class NotifyClient {
  private readonly baseUrl: string
  private readonly path: string
  private readonly token: string
  private readonly tokenHint: string
  private readonly fetchImpl: typeof globalThis.fetch
  private readonly timeoutMs: number
  private readonly logger: Logger

  constructor(cfg: NotifyConfig) {
    if (!cfg.baseUrl) {
      throw new Error(
        'octoauth/internal-helpers: NotifyConfig.baseUrl is required',
      )
    }
    if (!cfg.tokenEnvVar) {
      throw new Error(
        'octoauth/internal-helpers: NotifyConfig.tokenEnvVar is required',
      )
    }
    const token = process.env[cfg.tokenEnvVar]
    if (!token) {
      throw new Error(
        `octoauth/internal-helpers: env var "${cfg.tokenEnvVar}" is empty (fail-closed)`,
      )
    }
    this.baseUrl = cfg.baseUrl
    this.path = cfg.notifyPath ?? DEFAULT_NOTIFY_PATH
    this.token = token
    this.tokenHint = maskToken(token)
    if (cfg.fetch !== undefined) {
      this.fetchImpl = cfg.fetch
    } else {
      if (typeof globalThis.fetch !== 'function') {
        throw new Error(
          'octoauth/internal-helpers: globalThis.fetch is not available; pass NotifyConfig.fetch',
        )
      }
      this.fetchImpl = globalThis.fetch.bind(globalThis)
    }
    this.timeoutMs = cfg.timeoutMs ?? DEFAULT_NOTIFY_TIMEOUT_MS
    this.logger = cfg.logger ?? consoleLogger
  }

  /**
   * send posts req to the notify endpoint. Non-2xx status becomes an Error
   * whose message contains ONLY the status code — the response body is NOT
   * included so log scrapers cannot enumerate server-side messages via the
   * caller's error path.
   */
  async send(req: NotifyRequest, signal?: AbortSignal): Promise<void> {
    let body: string
    if (req.rawBody !== undefined && req.rawBody !== '') {
      body = req.rawBody
    } else {
      try {
        body = JSON.stringify({
          channel_id: req.channelId ?? '',
          message: req.message ?? '',
        })
      } catch (err) {
        throw new Error(
          `octoauth/internal-helpers: encode notify request: ${(err as Error).message}`,
        )
      }
    }

    const timeoutCtl = new AbortController()
    const timer = setTimeout(() => timeoutCtl.abort(), this.timeoutMs)
    const composite = signal
      ? AbortSignal.any([signal, timeoutCtl.signal])
      : timeoutCtl.signal
    let resp: Response
    try {
      resp = await this.fetchImpl(this.baseUrl + this.path, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Accept: 'application/json',
          'X-Internal-Token': this.token,
        },
        body,
        signal: composite,
      })
    } catch (err) {
      this.logger.debug('octoauth/internal-helpers: notify transport error', {
        endpoint: this.path,
        token_hint: this.tokenHint,
      })
      throw new Error(
        `octoauth/internal-helpers: notify request failed: ${(err as Error).message}`,
      )
    } finally {
      clearTimeout(timer)
    }
    // Drain the body so the connection can be reused.
    try {
      await resp.text()
    } catch {
      // Ignore drain errors.
    }
    this.logger.debug('octoauth/internal-helpers: notify sent', {
      endpoint: this.path,
      status: resp.status,
      token_hint: this.tokenHint,
    })
    if (resp.status < 200 || resp.status >= 300) {
      throw new Error(
        `octoauth/internal-helpers: notify failed: status ${resp.status}`,
      )
    }
  }
}

/**
 * newNotifyClient is a convenience factory function equivalent to
 * `new NotifyClient(cfg)`.
 */
export function newNotifyClient(cfg: NotifyConfig): NotifyClient {
  return new NotifyClient(cfg)
}

/**
 * maskToken returns a short prefix + "..." suitable for logs. Even 8
 * characters of an internal shared secret should not appear in application
 * logs verbatim.
 */
export function maskToken(t: string): string {
  if (t.length < 8) return '***'
  return t.slice(0, 8) + '...'
}
